log/DESIGN.md

```diff
@@ -0,0 +1,743 @@
+Additionally, `Severity[Level]1` constants are defined to make the API more readable and user friendly.
+
+```go
+type Severity int
+
+const (
+	SeverityTrace1 Severity = 1 // TRACE
+	SeverityTrace2 Severity = 2 // TRACE2
+	SeverityTrace3 Severity = 3 // TRACE3
+	SeverityTrace4 Severity = 4 // TRACE4
+
+	SeverityDebug1 Severity = 5 // DEBUG
+	SeverityDebug2 Severity = 6 // DEBUG2
+	SeverityDebug3 Severity = 7 // DEBUG3
+	SeverityDebug4 Severity = 8 // DEBUG4
+
+	SeverityInfo1 Severity = 9  // INFO
+	SeverityInfo2 Severity = 10 // INFO2
+	SeverityInfo3 Severity = 11 // INFO3
+	SeverityInfo4 Severity = 12 // INFO4
+
+	SeverityWarn1 Severity = 13 // WARN
+	SeverityWarn2 Severity = 14 // WARN2
+	SeverityWarn3 Severity = 15 // WARN3
+	SeverityWarn4 Severity = 16 // WARN4
+
+	SeverityError1 Severity = 17 // ERROR
+	SeverityError2 Severity = 18 // ERROR2
+	SeverityError3 Severity = 19 // ERROR3
+	SeverityError4 Severity = 20 // ERROR4
+
+	SeverityFatal1 Severity = 21 // FATAL
+	SeverityFatal2 Severity = 22 // FATAL2
+	SeverityFatal3 Severity = 23 // FATAL3
+	SeverityFatal4 Severity = 24 // FATAL4
+
+	SeverityTrace = SeverityTrace1
+	SeverityDebug = SeverityDebug1
+	SeverityInfo  = SeverityInfo1
+	SeverityWarn  = SeverityWarn1
+	SeverityError = SeverityError1
+	SeverityFatal = SeverityFatal1
+)
+```
+
+[`SeverityText`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext)
+is accessed using following methods:
+
+```go
+func (r *Record) SeverityText() string
+func (r *Record) SetSeverityText(s string)
+```
+
+[`Body`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+is accessed using following methods:
+
+```go
+func (r *Record) Body() Value
+func (r *Record) SetBody(v Value)
+```
+
+[Log record attributes](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes)
+are accessed using following methods:
+
+```go
+func (r *Record) WalkAttributes(f func(KeyValue) bool)
+func (r *Record) AddAttributes(attrs ...KeyValue)
+```
+
+`Record` has a `AttributesLen` method that returns
+the number of attributes to allow slice preallocation
+when converting records to a different representation:
+
+```go
+func (r *Record) AttributesLen() int
+```
+
+Additionally, `Record` has a `Clone` method facilities
+coping a record with no shared state (attributes "back" slice):
+
+```go
+func (r *Record) Clone() Record
+```
+
+The records attributes design and implemntation is based on
+[`slog.Record`](https://pkg.go.dev/log/slog#Record).
+It allows achieving high-performance access and manipulation of the attributes
+while keeping the API user friendly.
+It relieves the user from making his own improvements
+for reducing the number of allocations when passing attributes.
+
+The following defintions are implementing the abstractions
+described in [the specification](https://opentelemetry.io/docs/specs/otel/logs/#new-first-party-application-logs):
+
+```go
+type Value struct{}
```

Comment:
I don't see how `Value` can hold information if it's an empty struct.

Reply:
This is an implementation detail that I did not want to call out in the API design.

You can take a look at [the prototype](https://github.com/open-telemetry/opentelemetry-go/pull/4725) that the idea is to have the implementation similar to slog's: https://github.com/pellared/opentelemetry-go/blob/logs-design/log/value.go